tests/python/pants_test/backend/python/tasks/test_ctypes_integration.py

```diff
@@ -82,6 +83,56 @@ def test_ctypes_binary(self):
       binary_run_output = invoke_pex_for_output(pex)
       self.assertEqual(b'x=3, f(x)=17\n', binary_run_output)
 
+  def test_invalidation_ctypes(self):
+    """Test that the current version of a python_dist() is resolved after modifying its sources."""
+    with temporary_dir() as tmp_dir:
+      with self.mock_buildroot(
```

Comment:
Rather than mocking the buildroot, generally integration tests mock only the workdir while keeping the same buildroot.

That looks like `with self.temporary_workdir() as workdir:`, and no adjustment of the buildroot.

Reply:
Is this cool even when the tests modify testproject source files?

Reply:
Oh. Missed that.

If modifying files the mock buildroot is not a bad idea. There is at least one `@contextmanager` to assist with in-place-edit-and-then-restore usecases, but creating the mock buildroot is probably superior.

Thanks.

Reply:
(I think I may have introduced that contextmanager you're referring to and then removed it unless I'm mistaken -- the use of the `mock_buildroot` was in response to a comment @wisechengyi made on another PR of mine a while ago, I forget exactly which right now)